apps/rush-lib/src/cli/logic/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -147,6 +162,33 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
     }
 
     if (!packageDescription.dependencies.hasOwnProperty(dependencyName)) {
+      if (versionRange) {
+        // this means the current temp project doesn't provide this dependency,
+        // however, we may be able to use a different version. we prefer the latest version
+        const minimumVersion: string = '0.0.0';
+        let latestVersion: string = minimumVersion;
+
+        this.getTempProjectNames().forEach((otherTempProject: string) => {
+          const otherVersion: string | undefined =
+            this.tryEnsureDependencyVersion(dependencyName, otherTempProject, undefined);
```

Comment:
>undefined [](start = 78, length = 9)

why aren't we passing our versionRange into tryEnsureDependencyVersion?

Reply:
I see, if I understand right, when `versionRange = undefined` we do a lookup with no side effects, whereas when `versionRange` is defined then we "try to ensure" and possibly modify the shrinkwrap.  These semantics are confusing, especially when "tempProjectName" is undefined.

I would suggest to split it into two functions:  `getDependencyVersion()` is the non-destructive implementation which we would call here.  And `tryEnsureDependencyVersion()` would  do the updates, and always require a `versionRange`.

---
In reply to: [171011819](https://github.com/Microsoft/web-build-tools/pull/543#discussion_r171011819) [](ancestors = 171011819)